src/docs/arch-block.dox

```diff
@@ -1,9 +1,132 @@
+stream of chunks in the file (in reality, blocks from different checkpoints can
+be interleaved).
+
+@section block_usage How WiredTiger uses the block manager
+
+Whenever a page is read into the cache, WiredTiger will read its corresponding
+block from the data file. This block contains what's referred to as the
+"page image" in other subsystems and has the newest versions of data for each
+key.
+
+Conversely, when a page is reconciled whether that is due to a page eviction or
+by a checkpoint, a new page image is constructed and it is written as a new
+block in the data file. In the case of checkpoint, the newly written block is
+bound to the current checkpoint and in the case of eviction, it is associated
+with the next checkpoint.
+
+@section block_impl How the block manager is implemented
+
+We've already discussed the concept of a block and how the data file's on-disk
+representation is divided into smaller blocks that are each associated with a
+checkpoint.
+
+In this section, we'll dive into the file layout in more detail and talk about
+some of the core data structures used as part of block allocation.
+
+@subsection block_layout Block layout
+
+Each block consists of a page header, a block manager header followed by the
+actual page image.
+
+The page header contains some metadata about the page contents such as the type
+of page, whether it is compressed, whether it is encrypted, its write generation
+, etc.
+
+The block manager header is simpler. It contains the size of the following disk
+image as well as an optional checksum for when block-level checksumming is
+enabled.
+
+@subsection desc_block Descriptor block
+
+As discussed earlier, a file is divided up into blocks. The first block in a file
+is special as it contains metadata about the file and is referred to as the
+"descriptor block". It contains the WiredTiger major and minor version, a
+checksum of the block contents as well as a "magic" number to check against.
+
+The descriptor block serves as a safety check to ensure that the file being
+loaded into the block manager is actually a WiredTiger data file, that it
+belongs to a compatible version of WiredTiger and that the entire file has not
+been corrupted. WiredTiger also offers block-level checksums to defend against
+file corruption at a more granular level, which can be found in the
+\ref checksum section.
+
+@subsection block_lists Extent lists
+
+Internally, WiredTiger uses a data structure called an extent list or a
+\c WT_EXTLIST to track file usage. An extent list consists of a series of
+extents (or \c WT_EXT elements). Each extent tracks a file offset and size to
+track a portion of the file.
+
+There are three extent lists that are maintained per checkpoint:
+
+- \c alloc: The extents allocated for a given checkpoint.
+- \c avail: The extents that are unused and available for allocation.
+- \c discard: The extents freed in the current checkpoint. After the checkpoint
```

Comment:
@keithbostic I'm a bit hazy about this part. Is this how it works?

This is what makes sense to me but I wasn't able to see an obvious part of code where we do that.

Reply:
Yes, that's correct as best I recall. Those are the major groupings, certainly.